ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -2,60 +2,290 @@ import { drawOccupancyZonesTexts } from './drawOccupancyZonesTexts';
+    OCCUPANCY_ZONE_HEIGHT
+  );
+  ctx.fill();
+  ctx.stroke();
+};
+
+const ARROW_OFFSET_X = 1;
+const ARROW_OFFSET_Y = 1.5;
+const ARROW_WIDTH = 4.5;
+const ARROW_TOP_Y = 3.5;
+const ARROW_BOTTOM_Y = 6.5;
+
+const drawThroughTrain = ({
+  ctx,
+  arrivalTimePixel,
+}: Omit<DrawZone, 'departureTimePixel' | 'yPosition'>) => {
+  // Through trains are materialized by converging arrows like the following ones
+  //  ___
+  //  \_/
+  //  / \
+  //  ‾‾‾
+  ctx.beginPath();
+  // draw the upper part
+  ctx.moveTo(arrivalTimePixel - ARROW_OFFSET_X, OCCUPANCY_ZONE_Y_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTimePixel - ARROW_WIDTH, OCCUPANCY_ZONE_Y_START - ARROW_TOP_Y);
+  ctx.lineTo(arrivalTimePixel + ARROW_WIDTH, OCCUPANCY_ZONE_Y_START - ARROW_TOP_Y);
+  ctx.lineTo(arrivalTimePixel + ARROW_OFFSET_X, OCCUPANCY_ZONE_Y_START + ARROW_OFFSET_Y);
+  // draw the lower part
+  ctx.lineTo(arrivalTimePixel + ARROW_WIDTH, OCCUPANCY_ZONE_Y_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTimePixel - ARROW_WIDTH, OCCUPANCY_ZONE_Y_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTimePixel - ARROW_OFFSET_X, OCCUPANCY_ZONE_Y_START + ARROW_OFFSET_Y);
+  ctx.fill();
+  // draw the white separator in the middle
+  ctx.moveTo(arrivalTimePixel - ARROW_OFFSET_X, OCCUPANCY_ZONE_Y_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTimePixel + ARROW_OFFSET_X, OCCUPANCY_ZONE_Y_START + ARROW_OFFSET_Y);
+  ctx.stroke();
+};
+
+type DrawRemainingTrainsBox = {
+  ctx: CanvasRenderingContext2D;
+  remainingTrainsNb: number;
+  xPosition: number;
+};
+
+const drawRemainingTrainsBox = ({ ctx, remainingTrainsNb, xPosition }: DrawRemainingTrainsBox) => {
+  const textY = OCCUPANCY_ZONE_Y_START - REMAINING_TEXT_OFFSET;
+
+  ctx.fillStyle = REMAINING_TRAINS_BACKGROUND;
+  ctx.beginPath();
+  ctx.rect(xPosition, textY, REMAINING_TRAINS_WIDTH, REMAINING_TRAINS_HEIGHT);
+  ctx.fill();
+  ctx.stroke();
+  ctx.fillStyle = WHITE_100;
+  ctx.font = SANS;
+  ctx.textAlign = 'center';
+  ctx.textBaseline = 'middle';
+  ctx.fillText(
+    `+${remainingTrainsNb} trains`,
+    xPosition + REMAINING_TRAINS_WIDTH / 2,
+    textY + REMAINING_TRAINS_HEIGHT / 2
+  );
+};
+
+const drawOccupationZone = ({
+  ctx,
+  zone,
+  arrivalTimePixel,
+  departureTimePixel,
+  yPosition,
+  isThroughTrain,
+  selectedTrainId,
+}: {
+  ctx: CanvasRenderingContext2D;
+  zone: OccupancyZone;
+  arrivalTimePixel: number;
+  departureTimePixel: number;
+  yPosition: number;
+  isThroughTrain: boolean;
+  selectedTrainId: string;
+}) => {
+  ctx.fillStyle = zone.color;
+  ctx.strokeStyle = WHITE_100;
+  ctx.lineJoin = 'round';
+  ctx.lineCap = 'round';
+  ctx.font = '400 10px IBM Plex Mono';
+
+  const extraWidth = isThroughTrain ? X_TROUGHTRAIN_BACKGROUND_PADDING : X_BACKGROUND_PADDING;
+  const { originTextLength, destinationTextLength } = {
+    originTextLength: ctx.measureText(zone.originStation || '--').width,
+    destinationTextLength: ctx.measureText(zone.destinationStation || '--').width,
+  };
```

Comment:
I think the object is superflu here :)
```suggestion
  const originTextLength = ctx.measureText(zone.originStation || '--').width;
  const destinationTextLength = ctx.measureText(zone.destinationStation || '--').width;
```

Reply:
ok done